src/core/net/srp_advertising_proxy.cpp

```diff
@@ -0,0 +1,1393 @@
+    serviceInfo.mTxtDataLength       = aService.GetTxtDataLength();
+    serviceInfo.mPort                = aService.GetPort();
+    serviceInfo.mWeight              = aService.GetWeight();
+    serviceInfo.mPriority            = aService.GetPriority();
+    serviceInfo.mTtl                 = aService.GetTtl();
+    serviceInfo.mInfraIfIndex        = Get<BorderRouter::InfraIf>().GetIfIndex();
+
+    Get<Dnssd>().RegisterService(serviceInfo, aService.mAdvId, HandleRegistered);
+
+exit:
+    if (error != kErrorNone)
+    {
+        LogWarn("Error %s registering service '%s' '%s'", ErrorToString(error), aService.GetInstanceLabel(),
+                serviceName);
+    }
+}
+
+void AdvertisingProxy::UnregisterService(Service &aService)
+{
+    Dnssd::Service serviceInfo;
+    DnsName        hostName;
+    DnsName        serviceName;
+
+    aService.mShouldAdvertise = false;
+    aService.mIsRegistered    = false;
+    aService.mAdvId           = kInvalidRequestId;
+
+    CopyNameAndRemoveDomain(hostName, aService.GetHost().GetFullName());
+    CopyNameAndRemoveDomain(serviceName, aService.GetServiceName());
+
+    LogInfo("Unregistering service '%s' '%s' on '%s'", aService.GetInstanceLabel(), serviceName, hostName);
+
+    serviceInfo.Clear();
+    serviceInfo.mHostName        = hostName;
+    serviceInfo.mServiceInstance = aService.GetInstanceLabel();
+    serviceInfo.mServiceType     = serviceName;
+    serviceInfo.mInfraIfIndex    = Get<BorderRouter::InfraIf>().GetIfIndex();
+
+    Get<Dnssd>().UnregisterService(serviceInfo, 0, nullptr);
+}
+
+void AdvertisingProxy::RegisterKey(Host &aHost)
+{
+    DnsName hostName;
+
+    aHost.mShouldRegisterKey = false;
+
+    CopyNameAndRemoveDomain(hostName, aHost.GetFullName());
+
+    LogInfo("Registering key for host '%s', id:%lu", hostName, ToUlong(aHost.mKeyAdvId));
+
+    RegisterKey(hostName, /* aServiceType */ nullptr, aHost.mKey, aHost.mKeyAdvId, aHost.GetTtl());
+}
+
+void AdvertisingProxy::RegisterKey(Service &aService)
+{
+    DnsName serviceType;
+
+    aService.mShouldRegisterKey = false;
+
+    CopyNameAndRemoveDomain(serviceType, aService.GetServiceName());
+
+    LogInfo("Registering key for service '%s' '%s', id:%lu", aService.GetInstanceLabel(), serviceType,
+            ToUlong(aService.mKeyAdvId));
+
+    RegisterKey(aService.GetInstanceLabel(), serviceType, aService.mHost->mKey, aService.mKeyAdvId, aService.GetTtl());
+}
+
+void AdvertisingProxy::RegisterKey(const char      *aName,
+                                   const char      *aServiceType,
+                                   const Host::Key &aKey,
+                                   RequestId        aRequestId,
+                                   uint32_t         aTtl)
+{
+    Dnssd::Key             keyInfo;
+    Dns::Ecdsa256KeyRecord keyRecord;
+
+    keyRecord.Init();
+    keyRecord.SetFlags(Dns::KeyRecord::kAuthConfidPermitted, Dns::KeyRecord::kOwnerNonZone,
+                       Dns::KeyRecord::kSignatoryFlagGeneral);
```

Comment:
I just read [RFC3445 Section 3](https://www.rfc-editor.org/rfc/rfc3445.html#section-3) and found it's violating the following statement.
```
In the flags field, all bits except bit 7 are reserved and MUST be
   zero.  If Bit 7 (Zone bit) is set to 1, then the KEY is a DNS Zone
   key.  If Bit 7 is set to 0, the KEY is not a zone key.  SIG(0)/TKEY
   are examples of DNSSEC keys that are not zone keys.
```
Here `kSignatoryFlagGeneral` is 1 at the 16th bit.

I wonder if there's a following RFC that allows setting `1`s at bits other than the 7th bit?

Reply:
This is following the same flags and format as in key RR used by SRP (e.g. in SRP client or checked in SRP server). Here we are just re-constructing the key RR that SRP sever got from client.

Regarding registering key on mDNS in adv-proxy, its purpose is to keep the claim to them name the record data is not really important. There were discussions of even allowing empty record data for it.